src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.Object.cs

```diff
@@ -311,6 +311,187 @@ public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOp
             }
         }
 
+        private class PrimitiveConverter : JsonConverter<object>
+        {
+            public int ReadCallCount { get; private set; }
+            public int WriteCallCount { get; private set; }
+
+            public override bool CanConvert(Type typeToConvert)
+                => typeToConvert != typeof(ClassWithPrimitives)
+                && typeToConvert != typeof(ClassWithNullablePrimitives);
+
+            public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                ReadCallCount++;
+
+                if (reader.TokenType == JsonTokenType.True)
+                {
+                    return true;
+                }
+
+                if (reader.TokenType == JsonTokenType.False)
+                {
+                    return false;
+                }
+
+                if (reader.TokenType == JsonTokenType.Number)
+                {
+                    if (reader.TryGetInt32(out int i))
+                    {
+                        return i;
+                    }
+
+                    return reader.GetDouble();
+                }
+
+                if (reader.TokenType == JsonTokenType.String)
+                {
+                    if (reader.TryGetDateTime(out DateTime datetime))
+                    {
+                        return datetime;
+                    }
+
+                    return reader.GetString();
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+            {
+                WriteCallCount++;
+
+                if (value is int i)
+                {
+                    writer.WriteNumberValue(i);
+                }
+                else if (value is bool b)
+                {
+                    writer.WriteBooleanValue(b);
+                }
+                else if (value is string s)
+                {
+                    writer.WriteStringValue(s);
+                }
+                else
+                {
+                    throw new NotSupportedException();
+                }
+            }
+        }
+
+        private class ClassWithPrimitives
+        {
+            public int MyIntProperty { get; set; }
+            public bool MyBoolProperty { get; set; }
+            public string MyStringProperty { get; set; }
+#pragma warning disable 0649
+            public int MyIntField;
+            public bool MyBoolField;
+            public string MyStringField;
+#pragma warning restore
+        }
+
+        [Fact]
+        public static void ClassWithPrimitivesObjectConverter()
+        {
+            const string expected =
```

Comment:
Since reflection order is non-deterministic, we can't assume ordering of properties. Each property must be searched for independently or perhaps just a length check on the JSON.